pkg/logql/log/ip.go

```diff
@@ -0,0 +1,296 @@
+	default:
+		return nil, ErrIPFilterInvalidOperation
+	}
+
+	ip, err := newIPFilter(pattern)
+	if err != nil {
+		return nil, err
+	}
+	return &IPLineFilter{
+		ip: ip,
+		ty: ty,
+	}, nil
+}
+
+// Filter implement `Filterer` interface. Used by `LineFilter`
+func (f *IPLineFilter) Filter(line []byte) bool {
+	return f.filterTy(line, f.ty)
+}
+
+// ToStage implements `Filterer` interface.
+func (f *IPLineFilter) ToStage() Stage {
+	return f
+}
+
+// `Process` implements `Stage` interface
+func (f *IPLineFilter) Process(line []byte, _ *LabelsBuilder) ([]byte, bool) {
+	return line, f.filterTy(line, f.ty)
+}
+
+// `RequiredLabelNames` implements `Stage` interface
+func (f *IPLineFilter) RequiredLabelNames() []string {
+	return []string{} // empty for line filter
+}
+
+func (f *IPLineFilter) filterTy(line []byte, ty labels.MatchType) bool {
+	if ty == labels.MatchNotEqual {
+		return !f.ip.filter(line)
+	}
+	return f.ip.filter(line)
+}
+
+// `String` implements fmt.Stringer inteface, by which also implements `LabelFilterer` inteface.
+func (f *IPLineFilter) String() string {
+	// return fmt.Sprintf("%s%sip(%q)", ipf.labelName, eq, ipf.pattern) // label filter
+	return fmt.Sprintf("ip(%q)", f.ip.pattern)
+}
+
+type IPLabelFilter struct {
+	ip *ipFilter
+	ty LabelFilterType
+
+	// if used as label matcher, this holds the identifier label name.
+	// e.g: (|remote_addr = ip("xxx")). Here labelName is `remote_addr`
+	label string
+
+	// patError records if given pattern is invalid.
+	patError error
+}
+
+// NewIPLabelFilter is used to construct ip filter as label filter for the given `label`.
+func NewIPLabelFilter(pattern string, label string, ty LabelFilterType) *IPLabelFilter {
+	ip, err := newIPFilter(pattern)
+	return &IPLabelFilter{
+		ip:       ip,
+		label:    label,
+		ty:       ty,
+		patError: err,
+	}
+}
+
+// `Process` implements `Stage` interface
+func (f *IPLabelFilter) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+	return line, f.filterTy(line, f.ty, lbs)
+}
+
+// `RequiredLabelNames` implements `Stage` interface
+func (f *IPLabelFilter) RequiredLabelNames() []string {
+	return []string{f.label}
+}
+
+// PatternError will be used `labelFilter.Stage()` method so that, if the given pattern is wrong
+// it retuns proper 400 error to the client of LogQL.
+func (f *IPLabelFilter) PatternError() error {
+	return f.patError
+}
+
+func (f *IPLabelFilter) filterTy(line []byte, ty LabelFilterType, lbs *LabelsBuilder) bool {
+	if f.patError != nil {
+		lbs.SetErr(f.patError.Error())
+		return false
+	}
+
+	input, ok := lbs.Get(f.label)
+	if !ok {
+		lbs.SetErr(fmt.Sprintf("%s: %q %s", errLabelFilter, f.label, "label not found"))
+		return false
+	}
+
+	switch ty {
+	case LabelFilterEqual:
+		return f.ip.filter([]byte(input))
+	case LabelFilterNotEqual:
+		return !f.ip.filter([]byte(input))
+	}
+	// any other operation not supported
+	lbs.SetErr(fmt.Sprintf("%q, %s", ty, ErrIPFilterInvalidOperation))
```

Comment:
This should be validated too and not be there. Although I don't think it's possible. You might just want to remove it.